### PR TITLE
Add Second level Cache, Update Phpredis Client, Deprecate RedisCache, Fix session locking, Fix Travis

### PR DESCRIPTION
**Pull request by Pouix**
I saw that you were going to release a new version, so here is few things that I worked on:
- Add support for Doctrine Second Level Cache
- Rewrite PhpRedis client so it extends \Redis class
- Deprecate RedisCache class in favor of Doctrine\Common\Cache\RedisCache|PredisCache classes
- Fixed session locking: the lock can only be deleted by the one who put it
- Fixed Travis configuration for HHVM

Happy new year!

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -15,10 +15,10 @@
     "require": {
         "php": ">=5.3.3",
         "symfony/framework-bundle": "~2.1|~3.0",
-        "symfony/yaml": "~2.1|~3.0"
+        "symfony/yaml": "~2.1|~3.0",
+        "doctrine/cache": "1.*"
```
Comment (snc) [inline composer.json]: This is still optional, isn't it?

File: Client/Phpredis/Client.php [hunk 2]
```diff
@@ -113,4 +106,1296 @@ private function flatten($arguments, array &$list)
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function clearLastError()
+    {
+        return $this->call('clearLastError', func_get_args());
+    }
+
+    public function _prefix($value)
```
Comment (snc) [inline Client/Phpredis/Client.php]: Should be deleted like the other `_*` functions?!

**Comment by snc** [comment]
Thanks for your work, I added some line comments.

File: composer.json (hunk 1 above)
Comment (Pouix) [inline composer.json]: Indeed, in case we only use client or session parts.

File: Client/Phpredis/Client.php (hunk 2 above)
Comment (Pouix) [inline Client/Phpredis/Client.php]: Missed them.

**Comment by snc** [comment]
In the `RedisSessionHandler` class you added LUA code. Older redis versions do not support LUA. I think it would be OK to add a hint to the documentation mentioning the least required redis version, what do you think?

**Comment by Seldaek** [comment]
This needs to be rebased with latest master as well as it conflicts.

**Comment by Pouix** [comment]
@Seldaek It is rebased.

**Comment by Seldaek** [comment]
Thanks, however it's a bit too big for me to review safely right now, I'll let @snc decide if he wants to bring it in.

**Comment by snc** [comment]
I like the refactoring of the PhpRedis class to extend Redis. The only "problem" is see is that logging will not work for new commands (public methods). I'm not sure if a test could/should validate the overriden methods?

**Comment by Pouix** [comment]
@snc At first, I tried to use Reflection and generate the class, but Phpredis Redis class doesn't declare parameters of its public methods (except for `*scan` methods).
Maybe i'm wrong, but I didn't add a test for the same reason.

This is how I'm seeing it:
1. Manual refactoring of \Redis class from Phpredis (what is done with this PR).
2. Add the proper declarations on Phpredis with a PR (which I can't work on right now).
3. Generate and cache the class.

What do you think?

File: Resources/doc/index.md [hunk 3]
```diff
@@ -4,6 +4,10 @@
 
 This bundle integrates [Predis](https://github.com/nrk/predis) and [phpredis](https://github.com/nicolasff/phpredis) into your Symfony application.
 
+## Prerequisite ##
+
+Session handler requires `Redis >= 2.6`.
```
Comment (snc) [inline Resources/doc/index.md]: Could you please add "for LUA scripts" here?

**Comment by snc** [comment]
So when reflection currently can't be used, I'm OK to merge this PR as is.

File: Resources/doc/index.md (hunk 3 above)
Comment (Pouix) [inline Resources/doc/index.md]: This is fixed.

**Comment by snc** [comment]
Merged, big thanks.

**Comment by Pouix** [comment]
You're welcome.

**Comment by The-Don-Himself** [comment]
Is there a specific doctrine configuration for using the second_level_cache. I have been testing this out but keep getting exceptions such as 

````
Catchable Fatal Error: Argument 1 passed to Doctrine\Common\Cache\PredisCache::__construct() must implement interface Predis\ClientInterface, none given
````

**Comment by snc** [comment]
Looks like a bug, see #233.

**Comment by The-Don-Himself** [comment]
Ok, will test again once commit is merged and report accordingly.

**Comment by The-Don-Himself** [comment]
Can confirm that #233 fixes this. Queries are successfully cached and db updates automatically updates the cache as well.
Will have to move it to prod to test replication but think it should work flawlessly. Don't have a cluster step up though.
Funny thing though, I can't seem to see the cache ttl in the logs, doctrine second level has a 3600 default ttl with the previous method in given in #216 I would see this number in the logs but now I don't Will investigate more think its just an issue on my end.

**Pull request merged**